### PR TITLE
Database Refactor

### PR DESCRIPTION
**Pull request by pri2si17-1997**

### CODE REVIEW

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe @teryhill 

This covers following tables : 
1. patient_data
2. users
3. form_encounter
4. forms
5. All form_* tables excluding form_care_plan, form_functional_cognitive_status and form_observations.
6. facilities
7. amendments
8. amendment_history
9. calendar related table are not complete, it will be completed later on, just in case it can be required by other tables so I included with bare structure.

Please do check it, and let me know if any errors or improvements.

**Comment by aethelwulffe** [comment]
Hi Pri, this review may take me a day or two.  Can we get a .MD for the table structure?

**Comment by pri2si17-1997** [comment]
.MD means? Readme file? 

For database structure there is a .pdf file in `LaravelIntegration/Documentation/Tables` directory.

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe 

I am sorry for `Residential` i miss-spelled it. But regarding plural form of table name, that is typical laravel convention. This can be changed, but since there is valid reason of using plural forms in table name. 

https://laracasts.com/discuss/channels/eloquent/laravel-table-names-are-plural-by-default

A quick overview of tables can be obtained by looking at pdf file. 

I will certainly correct the mis-spelled word.

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe 

Anything extra needed?

**Comment by tmccormi** [comment]
*Secures* is not a valid English word regardless of any Laravel convention.   Database remapping is not really intended to be a laravel migration in anycase, just a cleanup and restructuring of the scheme.  No harm in using well known conventions, however.

**Comment by tmccormi** [comment]
I think what you need to do is run this so that the actual SQL is created and then dump the database so we can review the database structure.    Since we are not going to migrate LibreEHR to Laravel, all we need is the results of the work you have done to review.

File: LaravelIntegration/tables_to_be_dropped.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+1. array
+2. categories_seq
+3. documents_legal_detail
+4. documents_legal_categories
+5. documents_legal_master
+6. icd9_dx_code
+7. icd9_dx_long_code
+8. icd9_sg_code
+9. version
+10. sequences
```
Comment (teryhill) [inline LaravelIntegration/tables_to_be_dropped.txt]: Not sure you can drop this one with out changing some logic in the php files

**Comment by pri2si17-1997** [comment]
@tmccormi I agree that Secures is not a valid English Word, and I do used `UserSecure` as class name if you look at code, but it itself changed it to `user_secures`. I will change it if required.

Regarding structure, I have mentioned the instructions how to get that table imported. But I am uploading dump file as `LibreLaravel.sql` for review.

**Comment by aethelwulffe** [comment]
Anything that has the leeway to change itself (especially such an idiotic convention) gives me concerns...

**Comment by aethelwulffe** [comment]
OK, so I guess the way to test this is to take a db dump and use that as database.sql for a new install and see what happens, right?
  Aside from that, it seems like the Laravel stuff should perhaps be stored elsewhere (not in what would be production code) unless I am missing something important.  I really don't have a handle on this.

**Comment by pri2si17-1997** [comment]
Yes @aethelwulffe .

You can either test it through db dump file or get that migrated using laravel. I have added the instructions how to get that installed, but in case of any difficulty please let me know.

**Comment by aethelwulffe** [comment]
@pri2si17-1997  Not running a linux distro, and not sure I can take the time to play with Laravel at the moment.  I am pretty sure we are not going to add the laravel migration files to the base project immediately.  I think it should go into a different repo for optional modules or development/deployment stuff until we have made other significant changes and decisions.  I will try the dump file you added.

**Comment by pri2si17-1997** [comment]
Ok @aethelwulffe . No issues.

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe 

Have you looked at that dump file?

**Comment by aethelwulffe** [comment]
I don't even have an ethernet connection right now, and haven't since Saturday.  I know I am slacking, but I have been putting servers and networks back together after a weather event.  I should have a normal day tomorrow (I am at least on a network now) if I can get some other things fixed today.  Your sql dump is the very top of my list, rest assured.  My activity level is high right now, but it seems my performance is lacking in forward gear.  Sorry.

**Comment by aethelwulffe** [comment]
Well, the database loads from the dump (has a lot of lock table, unlock table stuff that I don't see the point of), but naturally, as the code does not match the table names ("s" added to some of them) the application does not run if I simply use the laravel dump and utilize it as sql/database.sql.
  The documentation is awesomely complete looking, despite that fact that I am still trying to get this: https://github.com/pri2si17-1997/LibreEHR/blob/GSoC_LibreEHR/LaravelIntegration/Documentation/install.txt running.  

I suppose I should be asking: if the Laravel stuff is all there from your branch as specified, does it do something magical that makes the tables all work?

I need some help here, because I am either correct in that the application needs extensive changes to accommodate the schema structure (not seeing that it has been done directly) or there is some magic Laravel stuff that is supposed to handle the migration for me.  I dunno.

**Comment by aethelwulffe** [comment]
If the status is "Art, we have the schema, and the intended next step is to update the code base to use it" I am ready to start helping out.

**Comment by aethelwulffe** [comment]
If you are saying:  "Lemme finish this stuff, then we can work on queries" then that sounds good to me!  Yes, finish the schema, *then* we do the other stuff.
Sounds like the only smart thing to me.
  Basically, you have not touched anything other than the Laravel Integration directory you created.  It is nice and clean this way (no conflicts) though I offer that you perhaps we should start a laravel repo that *just has your files*.

**Comment by pri2si17-1997** [comment]
@aethelwulffe 

> I offer that you perhaps we should start a laravel repo that just has your files.

I will be very happy to do this. But it is you guys to decide. :slightly_smiling_face: 
And what about laravel issues you faced?

**Comment by aethelwulffe** [comment]
@pri2si17-1997  I need some time (doing some heavy business stuff at the moment...couple of days), and I will be looking at the ICD-10 stuff to make sure we are covering the needs of the current code import loading, while not retaining any old useless stuff related to crosswalks to earlier versions or the like.

**Comment by pri2si17-1997** [comment]
Ok @aethelwulffe , No issues :slightly_smiling_face:  . I haven't changed anything in `ICD10*` structure as @teryhill told me. I though changed structure of few tables and need discussion with you. Waiting for you to get free, till then I'm working on queries, as after this I have to  move to that. I also figured out the working of `tf_patient_tags` :).

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe 

Please review the structure for `lists` and `documentations` table. I have dropped few columns. Also in `chart_tracker` table, column names were prefixed with `ct_`, so I removed that also.

In case of `lists`, we can store multiple diagnoses under diagnosis column. So I created a new table `list_diagnoses` to store diagnosis result. It will work as : 

**id | list_id | diagnosis**
1 | 1 | ICD10:E08.00
2 | 1 | ICD10:E08.329

It will prevent storing multiple value in one column.

**Comment by aethelwulffe** [comment]
[The form and tool repo EHR_Contribs is located here](https://github.com/LibreHealthIO/EHR_Contribs).  I would like to push all the laravel stuff there, then have you squash the actual database.sql output into a commit for the main repo and branch.
  I would also like to discuss with you @pri2si17-1997 the possibility of breaking up database.sql into a one-file-per-table set, then adjusting the setup tool to utilize those.  I am pretty sure this would make a LOT of things much more convenient.  We could, in each file, also add table version strings that could be parsed by another tool, or any number of other schemes.

**Comment by pri2si17-1997** [comment]
@aethelwulffe Can you please explain it a bit more? I did not get you.

**Comment by aethelwulffe** [comment]
@pri2si17-1997  What I am suggesting is that we push the LaravelIntegration directory to the indicated repo, and only push the actual database changes (and probably database documentation files) to the LibreEHR repo.  This keeps 230 new files that will not actually be doing anything for the end-user (from what I can see at the moment) out of the code-base footprint, and allow us to review and merge the Laravel integration project files in a different manner.  I assume that you have further ambitions with Laravel, but from the aspect of a singular clean commit, the implemented database.sql output from this project is what Tony indicated he was interested in first.
  Next stage should be making the LaravelIntegration stuff an optional module (for now...for people that can or are willing to install the laravel packages) that will live under LibreEHR/modules/LaravelIntegration.  I know your goal is to do MUCH more with this than merely clean up the database, and it very well may be that with some additional tweaks to installers (for windows) and other configuration scripts it will be valuable to use Laravel for ever more stuff, but we don't have project buy-in at the moment.  Let's make it a contrib repo option as a proper add-on module that can be dropped in, and if things grow from there, it thus becomes a core module, and so-on.
  I hope I am not confusing/disappointing you a bunch with all this, but it is merely a way to move forward cleanly, and allow folks to use laravel... but not *require* them to do so...at least for now.

**Comment by aethelwulffe** [comment]
@pri2si17-1997 @tmccormi  @teryhill @azwreith @nileshprasad137  am I way off base here?  Am I explaining myself well enough, or am I inherently missing some major concepts?
  Not having a frame or reference for all that the pieces of Priyanshu's LaravelIntegration directory is intended to do (looks like it is intended for doing database migrations and upgrades...which we have very much been discussing all along), my opinions and outlook is clouded by raw ignorance.

**Comment by pri2si17-1997** [comment]
@aethelwulffe 

> What I am suggesting is that we push the LaravelIntegration directory to the indicated repo, and only push the actual database changes (and probably database documentation files) to the LibreEHR repo.

Ok. I will do the same.

> Next stage should be making the LaravelIntegration stuff an optional module (for now...for people that can or are willing to install the laravel packages) that will live under LibreEHR/modules/LaravelIntegration.

I think this will be done later on, when we have everything clean.

> (looks like it is intended for doing database migrations and upgrades...which we have very much been discussing all along)

Yes, but database refactor is itself taking much time. It needs to have look at all queries and all tables that are interconnected. :disappointed:

**Comment by aethelwulffe** [comment]
Another note:
If someone were migrating from...say openemr 4.x to this version with LaravelMigration, they would first have to migrate to the CURRENT database version...then the Laravel tool would work.  As there are relatively few LibreEHR installations at this point in time, most every case for using LibreHealthEHR would be converting from an openemr installation, or it would be a new clean install.  With a clean install, we only need the database.sql file(s).  With an older database...we need a reference copy of the database scheme that the laravel migration scripts would be looking for; convert everything to that, then install the Laravel stuff, then run the migrations.

This is indeed a tough job Pri.  You are doing excellent work though!  One day you will get to work on something fun again! 😉 

I really can't merge this PR into master here.  WE CAN merge it into the contribs repo though!  THEN we can make small PR's and merge them quickly.  This is currently way too big to review easily.

**Comment by pri2si17-1997** [comment]
> I really can't merge this PR into master here. WE CAN merge it into the contribs repo though! THEN we can make small PR's and merge them quickly. This is currently way too big to review easily.

Ok @aethelwulffe I agree.

**Comment by tmccormi** [comment]
I agree with Art here, and yes, Art you are correct.  Short term we need the 'results' of the data upgrade in the mainline.   Later we can use the Laravel Integration parts for the enhance data migrations and other stuff.  I believe we are on the right track here.

File: LaravelIntegration/Documentation/Tables/LibreLaravel.sql [hunk 2]
```diff
@@ -2269,7 +2466,7 @@ CREATE TABLE `migrations` (
   `migration` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
   `batch` int(11) NOT NULL,
   PRIMARY KEY (`id`)
-) ENGINE=InnoDB AUTO_INCREMENT=1076 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+) ENGINE=InnoDB AUTO_INCREMENT=1084 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```
Comment (teryhill) [inline LaravelIntegration/Documentation/Tables/LibreLaravel.sql]: Just a general question @aethelwulffe  and @tmccormi  Should we be defaulting these.

File: LaravelIntegration/LibreLaravel/app/Http/Controllers/Auth/RegisterController.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * @param  array  $data
+     * @return \Illuminate\Contracts\Validation\Validator
+     */
+    protected function validator(array $data)
+    {
+        return Validator::make($data, [
+            'name' => 'required|string|max:255',
+            'email' => 'required|string|email|max:255|unique:users',
+            'password' => 'required|string|min:6|confirmed',
```
Comment (teryhill) [inline LaravelIntegration/LibreLaravel/app/Http/Controllers/Auth/RegisterController.php]: @aethelwulffe and @tmccormi should we be requiring 8 min?

Comment (aethelwulffe) [inline LaravelIntegration/LibreLaravel/app/Http/Controllers/Auth/RegisterController.php]: If you aren't keeping "admin/pass" for the demo, that's fine.  Personally, I would never use less than 12, and I prefer 16.

 See dog run.  <------<<<  12 characters

Comment (teryhill) [inline LaravelIntegration/LibreLaravel/app/Http/Controllers/Auth/RegisterController.php]: I personally think we should get away from admin and pass.

**Comment by aethelwulffe** [comment]
I was only fooling.  Setup and demo don't do admin/pass. Remember all 
that noise about "Pass Phrase"?  'See dog run.' is harder to crack than 
'&*# J7uPz'.  Especially because people would be too embarrassed to 
write down "See dog run." unless they were practicing writing from their 
textbook "Dick and Jane".  They had fun with that one at the BSides 
convention. Length (with human recognition) trumps complexity...and 
typing a long sentence is easier than a complex string.  I like pass 
phrases like "Up your *Trump* Hole:P" and other pleasant things.  I also 
like using code as passwords, like some one-line code in C for Euler's 
Totient Function.  You know, stuff that is easy to remember :)

Here is my favorite short password:

|:(){ :|: & };:|

|Go ahead.  Run that as a bash script.  It's a 14 character string that 
computers (and AV software) really don't seem to like.O:-)
|

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe , I have added seeders for patient tables. I have tried to keep the data as real as possible, like language and religions, but there may be some abrupt data. I have also maintained proper foreign key relationship. You please have a look at it.

**Comment by pri2si17-1997** [comment]
@aethelwulffe @teryhill 

Added seeder for user related tables. You please have a look at it.

**Comment by aethelwulffe** [comment]
Already did.
I will send you a link for the value scripts I use in my random data gen tool.  This can provide list of names for patients, insurance companies and all that.  May be of some use for you to generate random data on the fly, while preserving relationships and data continuity like having a heteronormal names for males and females and all that.

**Comment by pri2si17-1997** [comment]
Ok @aethelwulffe . Finally you replied... :relieved: I was thinking that I have done something wrong but now its fine. :slightly_smiling_face: 

Yeah you please give me that link. It may help me in more better way.

**Comment by aethelwulffe** [comment]
http://SunCoastConnection.com/Downloads/Printapatient_logic_source.zip
These are old badly formatted versions that saved as .txt, as I was printing it to take on a boat trip with me to go over while on a long sail.  I included some table dumps as comments for reference.  I kept it around for cases like this, though I am not sure what syntax recognition on your machine will think of it all.  Use a C-style syntax and most of it may make sense.

**Comment by aethelwulffe** [comment]
Good catch on the "factories" vs "facilities" naming thing!

**Comment by aethelwulffe** [comment]
NOTE:  in those files, the "Objects" file is mostly pseudo code.  Everything else is too, but it looks a lot more like C.

**Comment by pri2si17-1997** [comment]
@aethelwulffe I am not able to download from provided link. :disappointed:

**Comment by aethelwulffe** [comment]
Hm...  I just tried it.  Must be the firewall.
I will put it somewhere else.
One moment:

**Comment by pri2si17-1997** [comment]
ok.

**Comment by aethelwulffe** [comment]
http://librehealthehr.suncoastconnection.com/Downloads/PrintaPatient_A7_1.zip
http://librehealthehr.suncoastconnection.com/Downloads/Printapatient_logic_source.zip

This is a machine on a different network without so many restrictions.

**Comment by pri2si17-1997** [comment]
@aethelwulffe Do you have `PrintaPatient_A7_1.exe` for linux distro?

And for rest I will look at it, and discuss with you, probably in the evening(IST). :smile:

**Comment by aethelwulffe** [comment]
No, I don't have a lixux distro for that.  I could port it if I could learn to handle the file management properly.  It does run under wine/winetricks just fine I hear.

**Comment by pri2si17-1997** [comment]
Ok.. I haven't tried it on wine, but I'll look at it. No issues.

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe , here is screenshot from `x12_partners` table. You please check it if it needs any improvements from what you were having initially.

screenshot from 2017-08-15 02-22-06

**Comment by aethelwulffe** [comment]
Pretty much...and I will send you the sections of the x12 standard that apply to the ISA and GS segments that are the field names in here (they have minimum and maximum characters, defaults, etc...).
  -but I am going to do that after you have gotten some sleep.  Apparently, your clock says 02:30.  Also, it looks like you laptop needs charging anyway.  😈

**Comment by aethelwulffe** [comment]
Also,
  Get to a stopping point.  This GSOC thing is winding down.

**Comment by pri2si17-1997** [comment]
Yups... @aethelwulffe

**Comment by aethelwulffe** [comment]
Can we port it to a new repo with just the relevant code? (as an option).

**Comment by pri2si17-1997** [comment]
Hello @aethelwulffe @teryhill 

Can this be merged to https://github.com/LibreHealthIO/EHR_Contribs ?

**Comment by tmccormi** [comment]
I think we want to put this in it's very own repo.   I'll figure that out this week.

**Comment by Trodrige** [comment]
Hello, @pri2si17-1997 I see the great job you've been doing. I wish to work with you on this. I'm not clear on the work you've been doing, but I can be of great help with my Laravel skills.

**Comment by eran13com** [comment]
Hi, @pri2si17-1997  I am trying to work on the issue Move setup and upgrade to migration model #91.  I must say I am new to Laravel  so I am working the issue while learning. You seem to be pretty talented with Laravel and I like to team up with you. :)

**Comment by pri2si17-1997** [comment]
Hi @Trodrige @eran13com .

Regarding my work : Till now I refactored the database schema and created migrations for each of them. (Some modules are left like ACL and payments etc.).  @eran13com Well it will be a great fun learning laravel while working and I will like to work with both of you. :slightly_smiling_face: 

My further plan is to work on view but as I said in #91 , we need to discuss it with our mentors viz @aethelwulffe @tmccormi @teryhill .

**Comment by Trodrige** [comment]
Ok, @pri2si17-1997 I'm ready to start work. Indicate when you're ready, then @eran13com and I will get the task going.

**Comment by eran13com** [comment]
Hello @pri2si17-1997,
I am also ready to start. But as you said since we need to discuss this
matterwith our mentors we should arrange and discuss it.  😊

**Comment by aethelwulffe** [comment]
Hi guys,
This PR is going to remain up as the base mass of code for this project.  There is also a branch (with no PR) that has the tables broken up into files https://github.com/LibreHealthIO/LibreEHR/tree/database_breakup that may also be useful (might be useful for something).

**Comment by aethelwulffe** [comment]
@pri2si17-1997 I have this code stashed just in case you do not somehow.  In whatever case, I think you should push it somewhere safe, but librehealth official.  I am dropping the PR just as general cleanup of active issues.  -I DO HAVE IT STASHED.

**Pull request closed without merging**